diff -Nur proftpd-1.2.0pre3/contrib/xferstats.holger-preiss proftpd-1.2.0pre3.pld/contrib/xferstats.holger-preiss
--- proftpd-1.2.0pre3/contrib/xferstats.holger-preiss	Sun Oct 18 04:24:41 1998
+++ proftpd-1.2.0pre3.pld/contrib/xferstats.holger-preiss	Fri May 28 10:49:08 1999
@@ -1,4 +1,4 @@
-#! /usr/bin/perl
+#!/usr/bin/perl
 # ---------------------------------------------------------------------------
 #
 # USAGE: xferstats <options>
@@ -40,7 +40,7 @@
 # If you want to specify $mydom1 and $mydom2 manually you should edit the
 # next two lines to customize for your domain. This will allow your domain
 # to be separated in the domain listing.
-# $mydom1 = "debian";
+# $mydom1 = "pld";
 # $mydom2 = "org";
 
 # edit the next line to customize for your default log file
diff -Nur proftpd-1.2.0pre3/include/default_paths.h proftpd-1.2.0pre3.pld/include/default_paths.h
--- proftpd-1.2.0pre3/include/default_paths.h	Sun Oct 18 04:24:41 1998
+++ proftpd-1.2.0pre3.pld/include/default_paths.h	Fri May 28 10:47:56 1999
@@ -30,7 +30,7 @@
  * ftpcount, ftpwho, etc.
  */
 #ifndef RUN_DIR
-#define RUN_DIR			"/var/run/proftpd"
+#define RUN_DIR			"/var/run"
 #endif
 
 /* The location you wish to place any core files produced as a result of
@@ -51,7 +51,7 @@
  * overriden at runtime with the '-c' switch
  */
 #ifndef CONFIG_FILE_PATH
-#define	CONFIG_FILE_PATH	"/etc/proftpd.conf"
+#define	CONFIG_FILE_PATH	"/etc/ftpd/proftpd.conf"
 #endif
 
 /* The location of your `shells' file; a newline delimited list of
@@ -75,6 +75,6 @@
 /* Location of the file containing users that *cannot* use ftp
  * services (odd, eh?)
  */
-#define FTPUSERS_PATH		"/etc/ftpusers"
+#define FTPUSERS_PATH		"/etc/ftpd/ftpusers"
 
 #endif /* PROFTPD_PATHS_H */
--- proftpd-1.2.0rc2/configure.in~	Sun Sep 17 07:09:08 2000
+++ proftpd-1.2.0rc2/configure.in	Sun Sep 17 07:10:35 2000
@@ -994,13 +994,6 @@
 test "x$prefix" = xNONE && prefix=$ac_default_prefix
 test "x$exec_prefix" = xNONE && exec_prefix='${prefix}'
 
-AC_DEFINE_UNQUOTED(CONFIG_FILE_PATH,
-			"`eval echo "${sysconfdir}/proftpd.conf"`")
-AC_DEFINE_UNQUOTED(PID_FILE_PATH,
-			"`eval echo "${localstatedir}/proftpd.pid"`")
-AC_DEFINE_UNQUOTED(RUN_DIR,
-			"`eval echo "${localstatedir}/proftpd"`")
-
 prefix="$savedPrefix"
 exec_prefix="$savedExecPrefix"
 
