diff -ur proftpd-1.2.0pre4.orig/modules/mod_auth.c proftpd-1.2.0pre4/modules/mod_auth.c
--- proftpd-1.2.0pre4.orig/modules/mod_auth.c	Mon Aug 30 05:43:57 1999
+++ proftpd-1.2.0pre4/modules/mod_auth.c	Wed Sep  1 13:59:36 1999
@@ -525,7 +525,11 @@
   c = _auth_resolve_user(p,&user,&ourname,&anonname);
 
   if(!user) {
-    log_pri(LOG_NOTICE,"failed login from %s, '%s' is not a UserAlias.",inet_ascii(p,session.c->remote_ipaddr),origuser);
+    log_auth(LOG_NOTICE,"USER %s: user is not a UserAlias from %s [%s] to %s:%i",
+               user,session.c->remote_name,
+	       inet_ascii(p,session.c->remote_ipaddr),
+	       inet_ascii(p,session.c->local_ipaddr),
+	       session.c->local_port);
     goto auth_failure;
   }
 
@@ -533,7 +537,11 @@
   aclp = login_check_limits(main_server->conf,FALSE,TRUE,&i);
 
   if((pw = auth_getpwnam(p,user)) == NULL) {
-    log_pri(LOG_NOTICE,"failed login from %s, can't find user '%s'",inet_ascii(p,session.c->remote_ipaddr),user);
+    log_auth(LOG_NOTICE,"USER %s: no such user found from %s [%s] to %s:%i",
+               user,session.c->remote_name,
+	       inet_ascii(p,session.c->remote_ipaddr),
+	       inet_ascii(p,session.c->local_ipaddr),
+	       session.c->local_port);
     goto auth_failure;
   }
 
