diff -ur proftpd-1.2.0pre9.orig/modules/mod_auth.c proftpd-1.2.0pre9/modules/mod_auth.c
--- proftpd-1.2.0pre9.orig/modules/mod_auth.c	Sat Oct 23 07:18:49 1999
+++ proftpd-1.2.0pre9/modules/mod_auth.c	Thu Oct 28 15:18:57 1999
@@ -541,8 +541,11 @@
   c = _auth_resolve_user(p,&user,&ourname,&anonname);
 
   if(!user) {
-    log_pri(LOG_NOTICE, "USER %s (Login failed): User not a UserAlias.",
-	    origuser);
+    log_auth(LOG_NOTICE,"USER %s: user is not a UserAlias from %s [%s] to %s:%i",
+		origuser,session.c->remote_name,
+		inet_ascii(p,session.c->remote_ipaddr),
+		inet_ascii(p,session.c->local_ipaddr),
+		session.c->local_port);
     goto auth_failure;
   }
 
@@ -550,7 +553,11 @@
   aclp = login_check_limits(main_server->conf,FALSE,TRUE,&i);
 
   if((pw = auth_getpwnam(p,user)) == NULL) {
-    log_pri(LOG_NOTICE, "USER %s (Login failed): Can't find user.", user);
+    log_auth(LOG_NOTICE,"USER %s: no such user found from %s [%s] to %s:%i",
+		user,session.c->remote_name,
+		inet_ascii(p,session.c->remote_ipaddr),
+		inet_ascii(p,session.c->local_ipaddr),
+		session.c->local_port);
     goto auth_failure;
   }
 
