--- proftpd-1.2.0rc3/Makefile.in~	Tue Feb  6 15:29:08 2001
+++ proftpd-1.2.0rc3/Makefile.in	Tue Feb  6 16:13:08 2001
@@ -52,13 +52,13 @@
 	$(CC) $(LDFLAGS) -o $@ $(BUILD) $(LIBS)
 
 ftpcount: lib src
-	$(CC) $(LDFLAGS) -o $@ $(FTPCOUNT) $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ $(FTPCOUNT) -lsupp
 
 ftpwho: lib src
-	$(CC) $(LDFLAGS) -o $@ $(FTPWHO) $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ $(FTPWHO) -lsupp
 
 ftpshut: lib src
-	$(CC) $(LDFLAGS) -o $@ $(FTPSHUT) $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ $(FTPSHUT) -lsupp
 
 utils: ftpcount ftpwho ftpshut
 
