diff -ur proftpd-1.2.0rc1.orig/src/main.c proftpd-1.2.0rc1/src/main.c
--- proftpd-1.2.0rc1.orig/src/main.c	Wed Jul 12 20:16:38 2000
+++ proftpd-1.2.0rc1/src/main.c	Wed Jul 12 20:18:42 2000
@@ -981,9 +981,9 @@
     if(id && id->argc > 1)
       send_response("220","%s",(char*)id->argv[1]);
     else if(get_param_int(server->conf,"DeferWelcome",FALSE) == 1)
-      send_response("220", "ProFTPD " VERSION " Server ready.");
+      send_response("220", " Server ready.");
     else
-      send_response("220", "ProFTPD " VERSION " Server (%s) [%s]",
+      send_response("220", " Server (%s) [%s]",
            server->ServerName,server->ServerAddress);
   } else {
     send_response("220", "%s FTP server ready.", server->ServerAddress);

