diff -burNp proftpd-1.3.2.orig/src/log.c proftpd-1.3.2-dud/src/log.c
--- proftpd-1.3.2.orig/src/wtmp.c	2008-12-17 00:43:55.000000000 +0100
+++ proftpd-1.3.2-dud/src/wtmp.c	2009-02-09 14:24:34.271689801 +0100
@@ -129,9 +129,31 @@ int log_wtmp(char *line, const char *nam
   struct utmp ut;
   static int fd = -1;
 
+#if defined(HAVE_UTMPX_H)
+#if !defined(WTMPX_FILE) && defined(_PATH_WTMPX)
+# define WTMPX_FILE _PATH_WTMPX
+#endif
+  static char *wtmp_file = NULL;
+  if (wtmp_file == NULL) {
+    if (access(WTMPX_FILE"x", F_OK) == 0)
+      wtmp_file = WTMPX_FILE"x";
+    else if (access(WTMPX_FILE, F_OK) == 0)
+      wtmp_file = WTMPX_FILE;
+    else if (access(WTMP_FILE"x", F_OK) == 0)
+      wtmp_file = WTMP_FILE"x";
+    else if (access(WTMP_FILE, F_OK) == 0)
+      wtmp_file = WTMP_FILE;
+    else {
+      pr_log_pri(PR_LOG_WARNING, "cannot find "WTMPX_FILE);
+      return -1;
+    }
+  }
+#else
+  static char *wtmp_file = WTMP_FILE;
+#endif
   if (fd < 0 &&
-      (fd = open(WTMP_FILE, O_WRONLY|O_APPEND, 0)) < 0) {
-    pr_log_pri(PR_LOG_WARNING, "wtmp %s: %s", WTMP_FILE, strerror(errno));
+      (fd = open(wtmp_file, O_WRONLY|O_APPEND, 0)) < 0) {
+    pr_log_pri(PR_LOG_WARNING, "wtmp %s: %s", wtmp_file, strerror(errno));
     return -1;
   }
 
