diff -ur proftpd-1.2.2rc1.org/src/main.c proftpd-1.2.2rc1/src/main.c
--- proftpd-1.2.2rc1.org/src/main.c	Sun Apr  1 12:17:08 2001
+++ proftpd-1.2.2rc1/src/main.c	Sun Apr  1 12:27:24 2001
@@ -1024,9 +1024,9 @@
     if(id && id->argc > 1)
       send_response("220","%s",(char*)id->argv[1]);
     else if(get_param_int(server->conf,"DeferWelcome",FALSE) == 1)
-      send_response("220", "ProFTPD " VERSION " Server ready.");
+      send_response("220", " Server ready.");
     else
-      send_response("220", "ProFTPD " VERSION " Server (%s) [%s]",
+      send_response("220", " Server (%s) [%s]",	
            server->ServerName,serveraddress);
   } else {
     send_response("220", "%s FTP server ready.", serveraddress);
