--- proftpd-1.2.0rc2/modules/mod_xfer.c~	Sat Feb  3 00:57:02 2001
+++ proftpd-1.2.0rc2/modules/mod_xfer.c	Sat Feb  3 01:02:11 2001
@@ -201,7 +201,7 @@
 }
 
 #ifdef HAVE_SENDFILE
-static int _transmit_sendfile(int rate_bps, unsigned long count, off_t offset,
+static int _transmit_sendfile(int rate_bps, unsigned long count, off_t *offset,
 			       pr_sendfile_t *retval) {
   
   /* We don't use sendfile() if:
@@ -216,7 +216,7 @@
   }
 
  retry:
-  *retval = data_sendfile(retr_fd, &offset, session.xfer.file_size - count);
+  *retval = data_sendfile(retr_fd, offset, session.xfer.file_size - count);
 
   if(*retval == -1) {
     switch (errno) {
@@ -257,7 +257,7 @@
 }
 #endif /* HAVE_SENDFILE */
 
-static long _transmit_data(int rate_bps, unsigned long count, off_t offset,
+static long _transmit_data(int rate_bps, unsigned long count, off_t *offset,
 			   char *buf, long bufsize) {
 #ifdef HAVE_SENDFILE
   pr_sendfile_t retval;
@@ -813,7 +813,7 @@
         break;
       
       /* INSERT CODE HERE */
-      if((len = _transmit_data(rate_bps, cnt, respos, lbuf, bufsize)) == 0)
+      if((len = _transmit_data(rate_bps, cnt, &respos, lbuf, bufsize)) == 0)
 	break;
       
       if(len < 0) {
