--- proftpd-1.2.0pre3/contrib/xferstats.holger-preiss	Sun Oct 18 04:24:41 1998
+++ proftpd-1.2.0pre3.pld/contrib/xferstats.holger-preiss	Fri May 28 10:49:08 1999
@@ -1,4 +1,4 @@
-#! /usr/bin/perl
+#!/usr/bin/perl
 # ---------------------------------------------------------------------------
 #
 # USAGE: xferstats <options>
@@ -40,7 +40,7 @@
 # If you want to specify $mydom1 and $mydom2 manually you should edit the
 # next two lines to customize for your domain. This will allow your domain
 # to be separated in the domain listing.
-# $mydom1 = "debian";
+# $mydom1 = "pld";
 # $mydom2 = "org";
 
 # edit the next line to customize for your default log file
--- proftpd-1.3.0/include/default_paths.h~	2006-04-29 14:05:26.906490255 +0300
+++ proftpd-1.3.0/include/default_paths.h	2006-04-29 14:06:53.778425641 +0300
@@ -57,7 +57,7 @@
  * overriden at runtime with the '-c' switch
  */
 #ifndef PR_CONFIG_FILE_PATH
-# define PR_CONFIG_FILE_PATH	"/etc/proftpd.conf"
+# define PR_CONFIG_FILE_PATH	"/etc/ftpd/proftpd.conf"
 #endif
 
 /* The location of your `shells' file; a newline delimited list of
@@ -80,6 +80,6 @@
 /* Location of the file containing users that *cannot* use ftp
  * services (odd, eh?)
  */
-#define PR_FTPUSERS_PATH	"/etc/ftpusers"
+#define PR_FTPUSERS_PATH	"/etc/ftpd/ftpusers"
 
 #endif /* PROFTPD_PATHS_H */
--- proftpd-1.3.3b/configure.in~	2010-09-15 11:59:14.956175001 +0200
+++ proftpd-1.3.3b/configure.in	2010-09-15 12:05:20.368863997 +0200
@@ -2276,7 +2276,7 @@
 locale_dir="`eval echo ${locale_dir}`"
 AC_DEFINE_UNQUOTED(PR_LOCALE_DIR, "`eval echo "${locale_dir}"`")
 
-AC_DEFINE_UNQUOTED(PR_RUN_DIR, "`eval echo "${localstatedir}"`")
+AC_DEFINE_UNQUOTED(PR_RUN_DIR, "`eval echo "${localstatedir}/proftpd"`")
 AC_DEFINE_UNQUOTED(PR_CONFIG_FILE_PATH, "`eval echo "${sysconfdir}/proftpd.conf"`")
 AC_DEFINE_UNQUOTED(PR_PID_FILE_PATH, "`eval echo "${localstatedir}/proftpd.pid"`")
 
diff -ur proftpd-1.3.2e/Make.rules.in proftpd-1.3.3b/Make.rules.in
--- proftpd-1.3.2e/Make.rules.in	2008-11-19 04:51:38.000000000 +0100
+++ proftpd-1.3.3b/Make.rules.in	2009-11-18 22:48:25.000000000 +0100
@@ -11,6 +11,7 @@
 localedir=@localedir@
 localstatedir=@localstatedir@
 pkgconfigdir=@pkgconfigdir@
+rundir=@localstatedir@/proftpd
 mandir=@mandir@
 
 AR=@AR@
