diff -ruN proftpd-1.1.6pl2-old/config.h.in proftpd-1.1.6pl2/config.h.in
--- proftpd-1.1.6pl2-old/config.h.in	Mon Sep  7 21:15:42 1998
+++ proftpd-1.1.6pl2/config.h.in	Fri Sep 18 22:35:32 1998
@@ -69,7 +69,7 @@
 #undef ino_t
 
 /* Define to `mode_t' if <sys/types.h> doesn't define. */
-#undef umode_t
+/* #undef umode_t */
 
 /* Define if your system has _pw_stayopen variable (Irix specific?) */
 #undef HAVE__PW_STAYOPEN
