diff -ur proftpd-1.2.0pre9.orig/src/main.c proftpd-1.2.0pre9/src/main.c
--- proftpd-1.2.0pre9.orig/src/main.c	Mon Oct 11 06:28:50 1999
+++ proftpd-1.2.0pre9/src/main.c	Thu Oct 28 15:12:26 1999
@@ -640,7 +640,7 @@
   log_pri(pri, log);
   
   if(standalone && master)
-    log_pri(LOG_NOTICE, "ProFTPD %s standalone mode SHUTDOWN", VERSION);
+    log_pri(LOG_NOTICE, "ProFTPD %s standalone mode SHUTDOWN","PLD Linux");
   
   end_login(exitcode);
 }
@@ -944,9 +944,9 @@
     if(id && id->argc > 1)
       send_response("220","%s",(char*)id->argv[1]);
     else if(get_param_int(server->conf,"DeferWelcome",FALSE) == 1)
-      send_response("220", "ProFTPD " VERSION " Server ready.");
+      send_response("220", "ProFTPD " "PLD Linux" " Server ready.");
     else
-      send_response("220", "ProFTPD " VERSION " Server (%s) [%s]",
+      send_response("220", "ProFTPD " "PLD Linux" " Server (%s) [%s]",
            server->ServerName,server->ServerAddress);
   }
   else
@@ -1618,7 +1618,7 @@
     log_rm_run();
     PRIVS_RELINQUISH
     if(standalone)
-      log_pri(LOG_NOTICE,"ProFTPD %s standalone mode SHUTDOWN",VERSION);
+      log_pri(LOG_NOTICE,"ProFTPD %s standalone mode SHUTDOWN","PLD Linux");
   }
 
   _internal_abort();  
@@ -1943,7 +1943,7 @@
       addl_bindings(s);
     }
 
-  log_pri(LOG_NOTICE,"ProFTPD %s standalone mode STARTUP",VERSION);
+  log_pri(LOG_NOTICE,"ProFTPD %s standalone mode STARTUP","PLD Linux");
   server_loop();
 }
 
@@ -2098,7 +2098,7 @@
       break;
     */
     case 'v':
-      log_pri(LOG_NOTICE,"ProFTPD Version " VERSION);
+      log_pri(LOG_NOTICE,"ProFTPD Version " "PLD Linux");
       exit(0);
     case 'h':
       show_usage(0);
