--- ./include/proftpd.h.org	Mon Oct 22 01:32:03 2001
+++ ./include/proftpd.h	Mon Oct 22 01:33:00 2001
@@ -95,7 +95,7 @@
   volatile int sp_flags;		/* Session/Protection flags */
 
   p_in_addr_t data_addr;		/* Remote data address */
-  short data_port;			/* Remote data port */
+  unsigned short data_port;		/* Remote data port */
 
   unsigned char ident_lookups;		/* Is RFC931 (ident) protocol used? */
   char *ident_user;			/* User identified by ident protocol */
