diff -urN proftpd-1.2.0rc2.org/modules/mod_pam.c proftpd-1.2.0rc2/modules/mod_pam.c
--- proftpd-1.2.0rc2.org/modules/mod_pam.c	Sun Oct 22 14:19:54 2000
+++ proftpd-1.2.0rc2/modules/mod_pam.c	Sun Oct 22 14:20:16 2000
@@ -73,14 +73,14 @@
       /* PAM frees response and resp.  If you don't believe this, please read
        * the actual PAM RFCs as well as have a good look at libpam.
        */
-      response[i].resp = (pam_user && *pam_user) ? strdup(pam_user) : NULL;
+      response[i].resp = (pam_user) ? strdup(pam_user) : NULL;
       break;
       
     case PAM_PROMPT_ECHO_OFF:
       /* PAM frees response and resp.  If you don't believe this, please read
        * the actual PAM RFCs as well as have a good look at libpam.
        */
-      response[i].resp = (pam_pass && *pam_pass) ? strdup(pam_pass) : NULL;
+      response[i].resp = (pam_pass) ? strdup(pam_pass) : NULL;
       break;
       
     case PAM_TEXT_INFO:
