--- proftpd-1.2.0rc2/contrib/mod_ratio.c~	Sat Feb  3 19:33:21 2001
+++ proftpd-1.2.0rc2/contrib/mod_ratio.c	Sat Feb  3 19:35:45 2001
@@ -382,7 +380,7 @@
     {
       path = dir_realpath (cmd->tmp_pool, cmd->arg);
       if (path
-	  && dir_check (cmd->tmp_pool, cmd->argv[0], cmd->group, path, NULL)
+	  && dir_check (cmd->tmp_pool, cmd->argv[0], cmd->group, path)
 	  && fs_stat (path, &sbuf) > -1)
 	fsize = sbuf.st_size;
 
